File: engine/scripting/dotnet/src/RetroEngine/Scene/Scene2D.cs
// @file $Scene2D.cs.cs
//
// @copyright Copyright (c) 2026 Retro & Chill. All rights reserved.
// Licensed under the MIT License. See LICENSE file in the project root for full license information.
using RetroEngine.Core.Math;
using RetroEngine.Interop;

namespace RetroEngine.Scene;

public class Scene2D
{
    public static Scene2D Current { get; private set; } = new();

    internal int Generation { get; }

    private readonly Dictionary<EntityId, Entity> _instantiatedEntities = [];

    public Entity CreateNewEntity(in Transform transform = default)
    {
        var entityId = EntityExporter.CreateNewEntity(in transform);
        var entity = new Entity(this, entityId);
        _instantiatedEntities.Add(entityId, entity);
        return entity;
    }

    internal void RemoveEntity(EntityId id)
    {
        _instantiatedEntities.Remove(id);
    }
}

File: engine/runtime/src/scene/scene2d.cpp
/**
 * @file scene2d.cpp
 *
 * @copyright Copyright (c) 2026 Retro & Chill. All rights reserved.
 * Licensed under the MIT License. See LICENSE file in the project root for full license information.
 */
module retro.runtime;

namespace retro
{
    boost::optional<Entity &> Scene2D::get_entity(const EntityID id)
    {
        return entities_.get(id);
    }

    Entity &Scene2D::create_entity(const Transform &transform) noexcept
    {
        return entities_.emplace(transform);
    }

    void Scene2D::destroy_entity(const EntityID id)
    {
        entities_.remove(id);
    }

    boost::optional<Component &> Scene2D::get_component(const ComponentID id)
    {
        return components_.get(id).map([](const std::unique_ptr<Component> &component) -> Component &
                                       { return *component; });
    }

    void Scene2D::destroy_component(const ComponentID component_id)
    {
        components_.remove(component_id);
    }
} // namespace retro
